Sources/Basics/Sandbox.swift

```diff
@@ -13,23 +13,38 @@ import TSCBasic
 import TSCUtility
 
 public enum Sandbox {
+
+    /// Applies a sandbox invocation to the given command line (if the platform supports it),
+    /// and returns the modified command line. On platforms that don't support sandboxing, the
+    /// command line is returned unmodified.
+    ///
+    /// - Parameters:
+    ///   - command: The command line to sandbox (including executable as first argument)
+    ///   - strictness: The basic strictness level of the standbox.
+    ///   - writableDirectories: Paths under which writing should be allowed.
+    ///   - readOnlyDirectories: Paths under which writing should be denied, even if they would have otherwise been allowed by either the strictness level or paths in `writableDirectories`.
     public static func apply(
         command: [String],
+        strictness: Strictness = .default,
         writableDirectories: [AbsolutePath] = [],
-        strictness: Strictness = .default
+        readOnlyDirectories: [AbsolutePath] = []
     ) -> [String] {
         #if os(macOS)
-        let profile = macOSSandboxProfile(writableDirectories: writableDirectories, strictness: strictness)
+        let profile = macOSSandboxProfile(strictness: strictness, writableDirectories: writableDirectories, readOnlyDirectories: readOnlyDirectories)
         return ["/usr/bin/sandbox-exec", "-p", profile] + command
         #else
         // rdar://40235432, rdar://75636874 tracks implementing sandboxes for other platforms.
         return command
         #endif
     }
 
+    /// Basic strictness level of a sandbox applied to a command line.
     public enum Strictness: Equatable {
+        /// Blocks network access and all file system modifications.
         case `default`
+        /// More lenient restrictions than the default, for compatibility with SwiftPM manifests using a tools version older than 5.3.
         case manifest_pre_53 // backwards compatibility for manifests
+        /// Like `default`, but also makes temporary-files directories (such as `/tmp`) on the platform writable.
         case writableTemporaryDirectory
```

Comment:
would it better to have the call site pass the temp dirs as writable locations instead of defining a new strictness option for that? alternatively, should this be an argument on "default" strictness instead of it's own case?

Reply:
I had that (passing in the temporary directories from the client) initially actually, but that pushes the platform-specific nature of these paths onto the client.  In addition to the paths, there could actually be a varying number of them on different platforms.  But I agree that this shouldn't really be part of the strictness, but rather an option on the default strictness as you suggest.  I think there are more of those in the future, such as whether or not to allow network, etc.

So yes, in retrospect I think it was a mistake to add the temporary directory handling as a separate strictness level, and it's something to be unwound in the future.

Really what I'd like to do is to change the `Sandbox` struct to a `SandboxProfile` or some other name that suggests that it's a set of options, and then that profile could be configured once and passed around so it can be applied to multiple calls later without having to thread through all the individual parameters (and perhaps also cached using operating system calls based on the contents of that profile).  Then as we add more parameters, none of the intermediate code has to change.  The code that calls the subprocess could just apply the same profile to one command after another.  I'm hoping to get to that improvement soonish, in particular since I'm hopeful that it could allow compilation of the profile on macOS just once and the reuse of the compiled profile, saving time when starting sandboxed subprocesses.

At that point it would make sense to get rid of strictness, I think, and just have a set of individually controllable options, and then a convenience initializer to create a `SandboxProfile` that matches the SwiftPM 5.3 semantics.  That special case could then exist outside of the `SandboxProfile` code itself as an extension, which seems more correct than leaking it into the sandbox API.

All of course for a future PR, but hopefully not too long in the future.